Fix keep_size for arviz structures.

rpgoldman opened the pull request:
The `keep_size` kwarg for `sample_posterior_predictive` and `fast_sample_posterior_predictive` didn't handle `arviz.InferenceData` or `xarray.Dataset` arguments correctly when the `keep_size` option is chosen.

Addresses !4004

kyleabeauchamp commented:
I can confirm that this PR leads to a more reasonable output shape in my driver code.

For the purposes of code review, the linter changes in posterior_predictive.py lead to a pretty massive diff.  I wonder if it might make sense to do linting as a separate PR to ensure high visibility of math changes.

michaelosthege commented on pymc3/distributions/posterior_predictive.py:
```diff
@@ -224,7 +293,9 @@ def fast_sample_posterior_predictive(trace: Union[MultiTrace, Dataset, List[Dict
-            _ETPParent = UserDict  # this is not seen by mypy but will be executed at runtime.
+            _ETPParent = (
+                UserDict  # this is not seen by mypy but will be executed at runtime.
+            )
```
if the comment was placed before the line it refers to, the formatting would be a lot nicer.

rpgoldman replied:
done.

michaelosthege commented on pymc3/distributions/posterior_predictive.py:
```diff
@@ -234,36 +305,45 @@ def extend_trace(self, trace: Dict[str, np.ndarray]) -> None:
+            }
+        return (
+            ppc_trace.data
+        )  # this gets us a Dict[str, np.ndarray] instead of my wrapped equiv.
```
same formatting comment as above

rpgoldman replied:
done

michaelosthege commented on pymc3/sampling.py:
```diff
@@ -91,11 +93,13 @@
 
 
-def instantiate_steppers(model, steps, selected_steps, step_kwargs=None):
+def instantiate_steppers(_model, steps, selected_steps, step_kwargs=None):
```
docstring doesn't match.
I don't think there's a realistic path to smoothly change this API without breaking changes. Instead let's make sure that we break, fix & mention it prominently?

rpgoldman replied:
I'm not sure what to do about this, since I made this change to quash a lint warning and reflect the fact that `model` is bound  but not read.  I didn't change the API at all, just updated the docstring and changed the variable name to reflect existing practice.

If you'd prefer, I'm happy to just undo this, but I'm not sure why it would be a breaking change.

michaelosthege replied:
Oh I though you wanted to deprecate the argument in the long run?
I don't see how re-naming it helps. Also, theoretically, someone _could_ pass it as `instantiate_steppers(model=...)` and that would break.

rpgoldman replied:
I used the underscore to tell pylint not to complain about the unread variable...

rpgoldman replied:
If you prefer, I could put the name back and add a pylint disable comment

michaelosthege replied:
sounds fine

michaelosthege reviewed (approved):
Generally looks good, but I have to admin that I probably missed half of the changes because of black formatting 🙄.

michaelosthege reviewed (changes requested):
actually I wanted to request changes - see the thread on `instantiate_steppers`

rpgoldman commented:
> I can confirm that this PR leads to a more reasonable output shape in my driver code.
> 
> For the purposes of code review, the linter changes in posterior_predictive.py lead to a pretty massive diff. I wonder if it might make sense to do linting as a separate PR to ensure high visibility of math changes.

I'm not sure that's feasible, since the CI checks for conformance with black formatting. I think the problem is that somehow this file slipped in without being black formatted to begin with.

Another time I would probably make the black formatting a separate commit but I didn't expect such a major upset.  Also, even if it's a separate commit, people mostly just compare the end state of a MR, so that's probably extra work to no effect.

I wish I had a better answer for you.

rpgoldman commented:
I'm going to rebase and rewrite history to try to separate out the reformatting, please wait for more updates.

rpgoldman committed:
Fix posterior pred. sampling keep_size w/ arviz input.

Previously posterior predictive sampling functions did not properly
handle the `keep_size` keyword argument when getting an xarray Dataset
as parameter.

Also extended these functions to accept InferenceData object as input.

rpgoldman committed:
Reformatting.

rpgoldman commented:
@kyleabeauchamp @michaelosthege I have just (force) pushed a drastic restructuring that splits the MR into code changes and reformatting. We should make sure it passes tests again -- in case I messed something up in the rebase.

Now you should be able to quickly review the code changes and separately evaluate the reformatting -- if you even care about the latter.

I appreciate your patience.

michaelosthege commented:
@rpgoldman much nicer to read!

I didn't find a problem. Your changes are also covered by the tests - except for raising the error raised by passing the wrong argument type.

rpgoldman commented:
> I didn't find a problem. Your changes are also covered by the tests - except for raising the error raised by passing the wrong argument type.

Do you mean https://github.com/pymc-devs/pymc3/blob/de2d6149453251b9b2496627c8fadffe59a83873/pymc3/distributions/posterior_predictive.py#L210 ?  Should I supply a test with an argument of inappropriate type here?

If not, please LMK what error you had in mind. Thanks!

kyleabeauchamp commented:
I think this LGTM as well (good tests, improved typing, fixing target issue), but I'll defer to others for approval as I'm not very familiar with this part of the codebase.

rpgoldman committed:
Check type errors.

Make errors consistent across sample_posterior_predictive and fast_sample_posterior_predictive, and add 2 tests.

rpgoldman commented:
@michaelosthege Added the two tests. If this passes, and is satisfactory, please merge. Thanks!

michaelosthege commented:
Entry into release notes is missing.. I'm in a hurry and can't find the button to edit it online..

rpgoldman committed:
Add changelog entry.

rpgoldman commented:
> Entry into release notes is missing.. I'm in a hurry and can't find the button to edit it online..

Fixed.
